src/crypto/crypto_context.cc

```diff
@@ -222,6 +227,245 @@ unsigned long LoadCertsFromFile(  // NOLINT(runtime/int)
+           trust_settings_result == kSecTrustSettingsResultTrustAsRoot ?
+               TRUSTED : UNSPECIFIED;
+  }
+
+  return UNSPECIFIED;
+}
+
+enum TrustStatus IsTrustSettingsTrustedForPolicy(CFArrayRef trustSettings,
+                                            bool isSelfIssued) {
+  // The trustSettings parameter can return a valid but empty CFArrayRef.
+  // This empty trust-settings array means “always trust this certificate”
+  // with an overall trust setting for the certificate of
+  // kSecTrustSettingsResultTrustRoot
+  if (CFArrayGetCount(trustSettings) == 0) {
+    if (isSelfIssued) {
+      return TRUSTED;
+    }
+  }
+
+  CFIndex trustSettingsCount = CFArrayGetCount(trustSettings);
+
+  for (CFIndex i = 0; i < trustSettingsCount ; ++i) {
+    CFDictionaryRef trustDict = (CFDictionaryRef) CFArrayGetValueAtIndex(
+        trustSettings, i);
+
+    enum TrustStatus trust = IsTrustDictionaryTrustedForPolicy(trustDict);
+
+    if (trust == DISTRUSTED) {
+      return trust;
+    } else if (trust == TRUSTED) {
+      return trust;
+    }
+  }
+  return UNSPECIFIED;
+}
+
+bool IsCertificateTrustValid(SecCertificateRef ref) {
+  SecTrustRef secTrust = nullptr;
+  CFMutableArrayRef subjCerts = CFArrayCreateMutable(
+      nullptr, 1, &kCFTypeArrayCallBacks);
+  CFArraySetValueAtIndex(subjCerts, 0, ref);
+
+  SecPolicyRef policy = SecPolicyCreateBasicX509();
+  OSStatus ortn = SecTrustCreateWithCertificates(subjCerts, policy, &secTrust);
+  bool result = false;
+  if (ortn) {
+    /* should never happen */
+    goto done;
+  }
+
+  result = SecTrustEvaluateWithError(secTrust, nullptr);
+  done:
+    if (policy) {
+      CFRelease(policy);
+    }
+    if (secTrust) {
+      CFRelease(secTrust);
+    }
+    if (subjCerts) {
+      CFRelease(subjCerts);
+    }
+    return result;
+}
+
+bool IsCertificateTrustedForPolicy(X509* cert, SecCertificateRef ref) {
+  OSStatus err;
+
+  bool trustEvaluated = false;
+  bool isSelfSigned = IsSelfSigned(cert);
+
+  // Evaluate user trust domain, then admin. User settings can override
+  // admin (and both override the system domain, but we don't check that).
+  for (const auto& trust_domain :
+       {kSecTrustSettingsDomainUser, kSecTrustSettingsDomainAdmin}) {
+    CFArrayRef trustSettings;
+    err = SecTrustSettingsCopyTrustSettings(ref, trust_domain, &trustSettings);
+
+    if (err == errSecSuccess && trustSettings != nullptr) {
+      TrustStatus result = IsTrustSettingsTrustedForPolicy(
+          trustSettings, isSelfSigned);
+      if (result != UNSPECIFIED) {
+        CFRelease(trustSettings);
+        return result == TRUSTED;
+      }
+    }
+
+    // An empty trust settings array isn’t the same as no trust settings,
+    // where the trustSettings parameter returns NULL.
+    // No trust-settings array means
+    // “this certificate must be verifiable using a known trusted certificate”.
+    if (trustSettings == nullptr && !trustEvaluated) {
```

Comment:
Isn't this going against what the docs says? The certificate is only to be trusted as root certificate if is not null but points to an empty array? i.e. this should already be covered in IsTrustSettingsTrustedForPolicy?

Reply:
No is trust settings won’t find it as there’s no trust settings. The validate is implementing the part of the statement must be validated by another certificate. This is for intermediate certificates that are not sent by the web server but are present in the OS keychain

Reply:
In that case shouldn't this be removed from the global root certificate array? And verified using `SecPolicyCreateSSL` instead on a per-connection basis? IIUC this is taking all the intermediate certificates, doing a basic check with `SecPolicyCreateBasicX509`, and if it looks X509 compliant, add it to the global root certificate array, which seems unsafe..

Reply:
SecTrustEvaluateWithError is validating that this is a trusted certificate.

Potentially, I suspect that’s what Chromium is doing although I didn’t find the code, intermediate certificates do work there

Reply:
As far as I can tell chromium stores intermediates here:
https://github.com/chromium/chromium/blob/main/net/cert/internal/trust_store_mac.cc#L823

The code is quite hard to follow.

----

When evaluating client certificates they do use `SecTrustEvaluateWithError` in order to get the full chain:
https://github.com/chromium/chromium/blob/98f89988c9774d0e138a0724aa64c46187203a77/net/ssl/client_cert_store_mac.cc#L83-L84

----

I think this is compliant but let me know if you think there's a better way of doing it